### ISSUE TITLE
Performance issue in Indentation check

### ISSUE THREAD

**Issue by mkordas** [checkstyle/checkstyle#2627]
Method [com.puppycrawl.tools.checkstyle.checks.indentation.HandlerFactory#getHandler](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/HandlerFactory.java#L137) is consuming huge amount of resources:

![image](https://cloud.githubusercontent.com/assets/5467276/11254807/22988e76-8e43-11e5-9cc6-eedcba9bb42f.png)

I think it's because `createdHandlers` cache is not updated properly and each time new instance of handler is created using reflection.
##

**Comment by rnveach** [comment]
As the code is written now, `getHandler`'s cache, `createdHandlers`, is only updated for method calls, which also matches what it's javadoc says: `Cache for created method call handlers.`
This seems to be working properly during `IndentationCheckTest.testValidMethodWithChecker`.

I do question the purpose of this cache.
It caches based on the AST. Besides what I see for nested method calls, even if it was populated with everything, it's storage would never be used again as we don't examine the same AST twice.

@mkordas I assume you are expecting `getHandler` to return the same instance for similar statements? Ex: All `imports` statements should re-use the same Handler instance, while all `package` statements should re-use a different Handler instance.

That is the only way I see this method using less reflection.

**Comment by mkordas** [comment]
@rnveach yes, I just expect this method to be as fast as possible. I'm not deep enough into that code right now to propose something, but I definitely know that there's something wrong with the current solution. I think that as long as our tests are passing we are safe to modify that code.

**Comment by DakshRJain737** [comment]
> [@mkordas](https://github.com/mkordas) I assume you are expecting `getHandler` to return the same instance for similar statements? Ex: All `imports` statements should re-use the same Handler instance, while all `package` statements should re-use a different Handler instance.
> 
> That is the only way I see this method using less reflection, otherwise we require a big switch and hardcode all the handlers.

@rnveach Will the solution you suggested work? Because the handler for any `TokenType` depend on its ast and parentAst. And for each ast(suppose import) the handler will never used again as AST will only be formed once.

**Comment by romani** [comment]
We need to reach to proper stable behavior for this Check before we start cosidering performance optimizations like this, there still too much defect in this Check.
